src/Npgsql/Internal/TypeHandlers/TextHandler.cs

```diff
@@ -120,25 +121,31 @@ static async ValueTask<string> ReadLong(NpgsqlReadBuffer buf, int byteLen, bool
     async ValueTask<char> INpgsqlTypeHandler<char>.Read(NpgsqlReadBuffer buf, int len, bool async, FieldDescription? fieldDescription)
     {
         // Make sure we have enough bytes in the buffer for a single character
+        // We can get here a much bigger length in case it's a string
+        // while we want to read only its first character
         var maxBytes = Math.Min(buf.TextEncoding.GetMaxByteCount(1), len);
         await buf.Ensure(maxBytes, async);
 
-        return ReadCharCore();
+        var character = ReadCharCore();
```

Comment:
Not too sure why we even have ReadCharCore as an inner method - we usually do that pattern when the outer method is non-async, to avoid the async state machine when I/O isn't needed; but we do await above anyway, so we may as well just inline it?

(I really don't think this code path is worth optimizing to the extent that we have an optimized non-async path)

Reply:
We can't inline because of `stackalloc` :trollface:

Reply:
Oh right, that was the reason :sweat_smile: